Fix intan `one-file-per-channel`

h-mayorquin opened the pull request:
Fix #1599 

The fix is composed of two steps:
1) Use "stream_name" as a reference to the channel maps instead of the `stream_index`
2) Sort the file names of in the `one-file-per-channel` case based on their channel names.

The first point also has the advantage that it removes the need to keep yet another numeric reference (that is not stream_id or stream_index) for the different data structures.

zm711 commented:
Love the use of the dicts instead of the lists. I'm busy for the next couple days, but since this ~200 lines of change I'll try to carefully read this at some point this week so we can add this in. Unless Sam has the bandwidth before me :)

h-mayorquin commented:
Thanks, @zm711

zm711 reviewed (commented):
My only question is how do I know which stream index to use (ie if I as a previous user < 14.0 if this PR makes it....)? Like if I were stream_index 15 in my code previously would this be a breaking change? I can test if needed.

Otherwise I think this looks really great. A lot better--definitely more robust.

h-mayorquin commented on neo/rawio/intanrawio.py:
```diff
@@ -743,10 +741,19 @@ def read_rhs(filename, file_format: str):
+
+        # Build a dictionary with channel count
+        special_cases_for_counting = [
+            "DC Amplifier channel",  # Same as RHS2000 amplifier channel
```
Yes, this is an improvement.

h-mayorquin commented on neo/rawio/intanrawio.py:
```diff
@@ -743,10 +741,19 @@ def read_rhs(filename, file_format: str):
+            "Stim channel",  # Same as RHS2000 amplifier channel
+        ]
+        names_to_count = [name for name in stream_names if name not in special_cases_for_counting]
+        channel_number_dict = {name: len(stream_name_to_channel_info_list[name]) for name in names_to_count}
```
If you think that's better, let's do it.

h-mayorquin replied:
Done.

h-mayorquin commented on neo/rawio/intanrawio.py:
```diff
@@ -1278,25 +1268,27 @@ def create_one_file_per_channel_dict_rhd(dirname):
-
-    raw_file_paths_dict[6] = [Path(dirname / "time.dat")]
-
+    for stream_name, file_prefix in stream_name_to_file_prefix_rhd.items():
```
I like adding issues as provenance but I think the best place for them is in the regression test. That said, I don't see why we can't have it on both places for this example.

As a side comment, I think it would be even better if I could also put the channel files on a dict but that would require a bigger refactor than I had the time to do.

h-mayorquin commented on neo/rawio/intanrawio.py:
```diff
@@ -1317,15 +1309,19 @@ def create_one_file_per_channel_dict_rhs(
+    raw_file_paths_dict["timestamp"] = [Path(dirname / "time.dat")]
     # 10 and 11 are hardcoded in the rhs reader so hardcoded here
-    raw_file_paths_dict[10] = [file for file in files if "dc-" in file.name]
+    raw_file_paths_dict["DC Amplifier channel"] = [file for file in files if "dc-" in file.name]
```
Yes. Good point.

h-mayorquin commented on neo/rawio/intanrawio.py:
```diff
@@ -1317,15 +1309,19 @@ def create_one_file_per_channel_dict_rhs(
 
     # we need time to be the last value
-    raw_file_paths_dict[15] = [Path(dirname / "time.dat")]
+    raw_file_paths_dict["timestamp"] = [Path(dirname / "time.dat")]
```
Good catch.

zm711 commented:
> Thanks @zm711 . I implemented your feedback.

Thanks yeah. I agree the design is a little tricky since the `rawio` exposes `stream_index` at the public level. I guess no particular `stream_index` is guaranteed, but in the case of `Intan` we have used the `stream_index` has a fossilized thing for so long (and I would say spikeglx, openephy, neuralynx, intan, and plexon are the most common ios we have used) that even though we don't have an official contract it might be disruptive to make a public api level change just in this case.  Are you okay if we run this by Sam or the whole team before merging? 

I guess to summarize my problem is that we don't actually have enough constraints on how to build a rawio and so different ios were build differently. I think this is a great improvement overall and will reduce mistakes. Could you run  the intan test suite from spikeinterface on this branch and double check that it converts everything appropriately to make sure at the SI level we didn't make any assumptions :)

h-mayorquin commented:
> so long (and I would say spikeglx, openephy, neuralynx, intan, and plexon are the most common ios we have used) that even though we don't have an official contract it might be disruptive to make a public api level change just in this case. Are you okay if we run this by Sam or the whole team before merging?

Sure, but we can also just run the index on master right now and re-order the signal stream header so it matches what we had before. Could you check that? I can check the Spikeinterface but I think that over there we don't rely on stream index.

h-mayorquin commented:
In fact, let's do just the re-ordering and then we don't need to check anything else. I will take care of it.

h-mayorquin commented:
Oh, in fact we are and were already sorting so this does should keep that behavior:

https://github.com/NeuralEnsemble/python-neo/blob/25760377b944ce2856766db3e6c90a8612ef69d9/neo/rawio/intanrawio.py#L246-L265

I also double-checked the order of the streams on the signals header (which is what determines stream_index) and found no change between master and this PR. Here is the code and output:

```python
from pathlib import Path
import subprocess
from neo.rawio import IntanRawIO
from pprint import pprint
import time 

neo_test_dir = Path("/home/heberto/ephy_testing_data/intan")
intan_test_cases_folder_paths = [p for p in neo_test_dir.iterdir() if p.is_dir()]

for test_case_folder_path in intan_test_cases_folder_paths:
    info_rhd_path = test_case_folder_path / "info.rhd"
    info_rhs_path = test_case_folder_path / "info.rhs"
    
    if info_rhd_path.exists() or info_rhs_path.exists():
        filename = info_rhd_path if info_rhd_path.exists() else info_rhs_path
        print("--------------------")
        print(test_case_folder_path.name)
        print(filename.suffix)
        # Switch to the master branch
        
        subprocess.run(["git", "switch", "master"], check=True)
        time.sleep(1) # Just so the print works
        reader = IntanRawIO(filename=filename)
        reader.parse_header()
        pprint(reader.header["signal_streams"])
        
        subprocess.run(["git", "switch", "fix_intan_one_file_per_channel"], check=True)
        time.sleep(1) # Just so the print works

        reader = IntanRawIO(filename=filename)
        reader.parse_header()
        pprint(reader.header["signal_streams"])

--------------------
rhd_fpc_multistim_240514_082044
.rhd
Your branch is up to date with 'origin/master'.
Already on 'master'
array([('RHD2000 amplifier channel', '0', ''),
       ('RHD2000 auxiliary input channel', '1', ''),
       ('USB board ADC input channel', '3', ''),
       ('USB board digital input channel', '4', ''),
       ('USB board digital output channel', '5', '')],
      dtype=[('name', '<U64'), ('id', '<U64'), ('buffer_id', '<U64')])
Your branch is up to date with 'origin/fix_intan_one_file_per_channel'.
Switched to branch 'fix_intan_one_file_per_channel'
array([('RHD2000 amplifier channel', '0', ''),
       ('RHD2000 auxiliary input channel', '1', ''),
       ('USB board ADC input channel', '3', ''),
       ('USB board digital input channel', '4', ''),
       ('USB board digital output channel', '5', '')],
      dtype=[('name', '<U64'), ('id', '<U64'), ('buffer_id', '<U64')])
--------------------
intan_fps_test_231117_052500
.rhd
Your branch is up to date with 'origin/master'.
Switched to branch 'master'
array([('RHD2000 amplifier channel', '0', '0'),
       ('RHD2000 auxiliary input channel', '1', '1'),
       ('USB board ADC input channel', '3', '3'),
       ('USB board digital input channel', '4', '4')],
      dtype=[('name', '<U64'), ('id', '<U64'), ('buffer_id', '<U64')])
Your branch is up to date with 'origin/fix_intan_one_file_per_channel'.
Switched to branch 'fix_intan_one_file_per_channel'
array([('RHD2000 amplifier channel', '0', '0'),
       ('RHD2000 auxiliary input channel', '1', '1'),
       ('USB board ADC input channel', '3', '3'),
       ('USB board digital input channel', '4', '4')],
      dtype=[('name', '<U64'), ('id', '<U64'), ('buffer_id', '<U64')])
--------------------
intan_fpc_test_231117_052630
.rhd
Your branch is up to date with 'origin/master'.
Switched to branch 'master'
array([('RHD2000 amplifier channel', '0', ''),
       ('RHD2000 auxiliary input channel', '1', ''),
       ('USB board ADC input channel', '3', ''),
       ('USB board digital input channel', '4', '')],
      dtype=[('name', '<U64'), ('id', '<U64'), ('buffer_id', '<U64')])
Your branch is up to date with 'origin/fix_intan_one_file_per_channel'.
Switched to branch 'fix_intan_one_file_per_channel'
array([('RHD2000 amplifier channel', '0', ''),
       ('RHD2000 auxiliary input channel', '1', ''),
       ('USB board ADC input channel', '3', ''),
       ('USB board digital input channel', '4', '')],
      dtype=[('name', '<U64'), ('id', '<U64'), ('buffer_id', '<U64')])
--------------------
intan_fps_rhs_test_240329_091536
.rhs
Your branch is up to date with 'origin/master'.
Switched to branch 'master'
array([('RHS2000 amplifier channel', '0', '0'),
       ('USB board ADC input channel', '3', '3'),
       ('USB board ADC output channel', '4', '4'),
       ('USB board digital input channel', '5', '5'),
       ('USB board digital output channel', '6', '6'),
       ('DC Amplifier channel', '10', '10'), ('Stim channel', '11', '11')],
      dtype=[('name', '<U64'), ('id', '<U64'), ('buffer_id', '<U64')])
Your branch is up to date with 'origin/fix_intan_one_file_per_channel'.
Switched to branch 'fix_intan_one_file_per_channel'
array([('RHS2000 amplifier channel', '0', '0'),
       ('USB board ADC input channel', '3', '3'),
       ('USB board ADC output channel', '4', '4'),
       ('USB board digital input channel', '5', '5'),
       ('USB board digital output channel', '6', '6'),
       ('DC Amplifier channel', '10', '10'), ('Stim channel', '11', '11')],
      dtype=[('name', '<U64'), ('id', '<U64'), ('buffer_id', '<U64')])
--------------------
intan_fpc_rhs_test_240329_091637
.rhs
Your branch is up to date with 'origin/master'.
Switched to branch 'master'
/home/heberto/development/python-neo/neo/rawio/intanrawio.py:826: UserWarning: Stim not implemented for `one-file-per-channel` due to lack of test files
  warnings.warn("Stim not implemented for `one-file-per-channel` due to lack of test files")
Switched to branch 'fix_intan_one_file_per_channel'
array([('RHS2000 amplifier channel', '0', ''),
       ('USB board ADC input channel', '3', ''),
       ('USB board ADC output channel', '4', ''),
       ('USB board digital input channel', '5', ''),
       ('USB board digital output channel', '6', ''),
       ('DC Amplifier channel', '10', '')],
      dtype=[('name', '<U64'), ('id', '<U64'), ('buffer_id', '<U64')])
Your branch is up to date with 'origin/fix_intan_one_file_per_channel'.
array([('RHS2000 amplifier channel', '0', ''),
       ('USB board ADC input channel', '3', ''),
       ('USB board ADC output channel', '4', ''),
       ('USB board digital input channel', '5', ''),
       ('USB board digital output channel', '6', ''),
       ('DC Amplifier channel', '10', '')],
      dtype=[('name', '<U64'), ('id', '<U64'), ('buffer_id', '<U64')])
/home/heberto/development/python-neo/neo/rawio/intanrawio.py:826: UserWarning: Stim not implemented for `one-file-per-channel` due to lack of test files
  warnings.warn("Stim not implemented for `one-file-per-channel` due to lack of test files")
```

zm711 reviewed (approved):
I'm good with this then. I'll ping Sam to look at our discussion, but as long as this maintains behavior then the discussion is more for future development than this PR. Thanks Heberto for putting up with me :)

h-mayorquin commented:
I branched the index discussion as a separate issue because it seems like an important thing:

https://github.com/NeuralEnsemble/python-neo/issues/1604

zm711 commented on neo/test/rawiotest/test_intanrawio.py:
```diff
@@ -66,6 +67,24 @@ def test_annotations(self):
+        folder_path = file_path.parent
+        amplifier_file_paths = [p for p in folder_path.iterdir() if "amp" in p.name]
+        
+        channel_names = [p.name[4:-4] for p in amplifier_file_paths]
+
+        for channel_name, amplifier_file_path in zip(channel_names, amplifier_file_paths):
+            data_raw = np.fromfile(amplifier_file_path, dtype=np.int16).squeeze()
+            data_from_neo = intan_reader.get_analogsignal_chunk(channel_ids=[channel_name], stream_index=0).squeeze()
+            np.testing.assert_allclose(data_raw, data_from_neo)
```
Could we get a note in comment about why this works...

so p.name[4:-4] means nothing to someone who doesn't know the file format. So I would explain why 4:-4. And if you're feeling ambitious maybe a better name than p (since I know a certain developer who hates one letter variable names....).

h-mayorquin replied:
Good idea. I was lazy here.

zm711 reviewed (commented):
One last comment then ready to merge @h-mayorquin .

zm711 commented:
Sorry I merged main into the branch, but this bug that is fixed on main is now popping up on RTD here. It appeared on the PR we just merged. Sorry let me try to track it down.

h-mayorquin commented:
git log revert?

zm711 commented:
It broke your other PR too. So I would love to forget it out. You want to look at the log and see if you see anything ? To me it looks like the code should be fine, so I don't know why it is using an old version of the plexon code.... I've been googling to see if there is a cache I can empty but seems like RTD doesn't make it easy.

zm711 commented:
Okay I found another Plexon overflow that looked the same as the last one so I didn't even realize it was a different one which had messed with me. Once we test and merge #1612 we can merge main into your PRs and that should fix everything :)

h-mayorquin commented:
I guessed that Pleon will bite you, one way or the other.

samuelgarcia commented:
Hi Ramon
This is a big clean. thank you.
OK for dict approach.
